### DISCUSSION TITLE
How to have the designer components send authentication tokens (cookies or bearer)

### DISCUSSION THREAD

**Discussion by masterpoi** [elsa-workflows/elsa-core#1161] [Q&A] [answered]
I've embedded the elsa-workflow-definition-editor-screen in my applications and it works fine. 
I used this code in a script tag to bootstrap the component: 

```javascript
import {defineCustomElements} from 'https://unpkg.com/@elsa-workflows/elsa-workflows-studio@2.0.1/dist/esm/loader.js';
defineCustomElements();
```

Now i'm trying to secure the communication between the frontend and the server by using identity server. Server side there is no issue since you can configure the aspnetcore pipeline as usual.

However, i need to configure the client code to send either a JWT token or a cookie to the backend. I guess i need to hook into the axios configuration passed in here https://github.com/elsa-workflows/elsa-core/blob/7763c88273a48cadd7478fe336e662f9547d3b03/src/designer/elsa-workflows-studio/src/services/elsa-client.ts#L25 .But i don't see any way to override those settings.

I've tried looking into setting axios global defaults, but i'm not too familiar with stencil and looking at 
@elsa-workflows/elsa-workflows-studio/dist/esm/elsa-client-41896e0a.js in the package, i don't see any way to access either the relevant axios object or the  elsa client.

Any suggestions on how to make the client send cookies or JWT/Bearer tokens?

P.s. I must say i love the project. Excellent work.

**Comment by sfmskywalker** [comment]
Thanks! 😄 

This is indeed a missing extension point that we should provide. I created an issue for it [here](https://github.com/elsa-workflows/elsa-core/issues/1164).

Once that's in place, you should be able to handle an event bus message called e.g. `'http-config-created'` by doing:

```typescript
import { eventBus } from '?/<-- path to import to be figured out';

eventBus.on('http-config-created', config => config.headers = ({'authorization': 'Bearer abc123'}));
```

**Comment by sfmskywalker** [comment]
Alternatively, we might just expose a simple DOM event on the `<elsa-workflow-definition-editor-screen>` element (as well as the lower-level designer element for folks who use that one directly). Then no need to figure out how to import the event bus service, keeping it even simpler:

```html
<elsa-workflow-definition-editor-screen onHttpConfigCreated="ConfigureHttpConfig"></elsa-workflow-definition-editor-screen>

<script>
function ConfigureHttpConfig(config) {
   config.headers = ({'authorization': 'Bearer abc123'};
}
</script>
```

**Reply by jdevillard** [reply]
this will be perfect, I've work on a similar solution with a DOM Event that allow the upper level to send event to webComponent to add middleware to axios.

The idea behing this was to be sure that the middleware (defined outside) was used for each call and the middleware allow also to customize easily how to manage the configuration depending on (for example) the URI.

Maybe it could be overkill... 

So my question is : is the onHttpConfigCreated event will be fired once or for each call? (because the header auth has to be refreshed)

**Reply by masterpoi** [reply]
I agree with @jdevillard. This looks like the perfect solution since it will integrate well everywhere.  Also, the axios docs mention interceptors that can be used to change the headers on each request: https://axios-http.com/docs/interceptors . We just need access to the axios object too, not only the config.

**Reply by sfmskywalker** [reply]
Makes sense. Although currently, the components will always create a new Axios instance making it sufficient to just configure the config object, this implementation detail may change in the future where the client is created once. So allowing access to the Axios instance or at least by allowing the installment of middleware / interceptors is a must.

**Comment by sfmskywalker** [comment, marked as answer]
You can now do so by adding a listener to the `httpClientCreated` event, which provides access to axios the middleware pipeline which lets you install additional middleware (like tacking on auth tokens to outgoing requests).

Example:

```html
<elsa-studio-root server-url="https://localhost:11000" monaco-lib-path="build/assets/js/monaco-editor/min"></elsa-studio-root>

<script>
    const elsaStudioRoot = document.querySelector('elsa-studio-root');
    elsaStudioRoot.addEventListener('httpClientCreated', event => {

        // Get access to the axios middleware service and register a sample middleware:
        const {service} = event.detail;

        // Register a sample middleware.
        service.register({
            onRequest(request) {
                console.log('onRequest');
                request.headers = {
                    ...request.headers,
                    'X-Client': 'Elsa Dashboard',
                    Authorization: 'Bearer secret-token' // For example, set a JWT value here.
                };
                return request;
            },

            onResponse(response) {
                console.log('onResponse');
                return response;
            }
        });

    })
</script>
```

**Reply by jdevillard** [reply]
Hum, 

I'm trying to use this, work well with the elsa-studio-root element but if we use only the designer screen , the external event are  not available. How can we access this  ?

**Reply by sfmskywalker** [reply]
You're right, that doesn't seem to work. I'll look into it.

**Reply by sfmskywalker** [reply]
Ah, I remember. The events are actually exposed via the new `<elsa-external-events>` component, which I added to the `<elsa-studio-root` component. 

I could add it to various high-level "screen" components such as the workflow definition editor or even lower-level designer, but there are bound to be other useful events coming from other components the host might want to handle. To prevent events being published more than once, I settled on requiring the host to add the `<elsa-external-events>` element to their page if they are not using the `<elsa-studio-root>` component.

It's not the most elegant solution, so I would be happy to revise this if someone has a clever suggestion.

**Reply by sfmskywalker** [reply]
In any case, the following works:

```html
<elsa-workflow-definition-editor-screen server-url="https://localhost:11000" monaco-lib-path="build/assets/js/monaco-editor/min"></elsa-workflow-definition-editor-screen>
<elsa-external-events></elsa-external-events>

<script>
    const elsaEvents= document.querySelector('elsa-external-events');
    elsaEvents.addEventListener('httpClientCreated', event => {

        // Get access to the axios middleware service and register a sample middleware:
        const {service} = event.detail;

        // Register a sample middleware.
        service.register({
            onRequest(request) {
                console.log('onRequest');
                request.headers = {
                    ...request.headers,
                    'X-Client': 'Elsa Dashboard',
                    Authorization: 'Bearer secret-token' // For example, set a JWT value here.
                };
                return request;
            },

            onResponse(response) {
                console.log('onResponse');
                return response;
            }
        });

    })
</script>

```

**Reply by jdevillard** [reply]
Ho yes ! Thanks,

I've test document.querySelector('elsa-workflow-definition-editor-screen') but not document.querySelector('elsa-external-events');

Seems to works ! now will try to embed the <script> tag in a more Angular compliant way ;)

**Reply by jdevillard** [reply]
ok you can do it : 
on your html : 
```
<elsa-external-events (httpClientCreated)="onHttpClientSaved($event)"></elsa-external-events>
```

and on you controller :  
```
 onHttpClientSaved(event: any)
  {
    const {service} = event.detail;

    // Register a sample middleware.
    service.register({
        onRequest(request) {
            console.log('onRequest');
            request.headers = {
                ...request.headers,
                'X-Client': 'Elsa Dashboard',
                Authorization: 'Bearer secret-token' // For example, set a JWT value here.
            };
            return request;
        },

        onResponse(response) {
            console.log('onResponse');
            return response;
        }
    });
  }
```

**Reply by sfmskywalker** [reply]
Neat!

**Reply by masterpoi** [reply]
Awesome. Will try it later too.

**Reply by mayaoxiansheng** [reply]
npm i @elsa-workflows/elsa-workflows-studio
elsa-external-events show unregistered，Can it be used like this?

**Reply by sfmskywalker** [reply]
@mayaoxiansheng 

Yes, sorry: the `<elsa-external-events>` has been removed. Instead, you now have to do this:

```html

<elsa-studio-root server-url="https://localhost:11000" monaco-lib-path="build/assets/js/monaco-editor/min">
   <elsa-workflow-definition-editor-screen ></elsa-workflow-definition-editor-screen>
</elsa-studio-root server-url>

<script>
    const elsaStudioRoot= document.querySelector('elsa-studio-root');
    elsaStudioRoot.addEventListener('httpClientCreated', event => {

        // Get access to the axios middleware service and register a sample middleware:
        const {service} = event.detail;

        // Register a sample middleware.
        service.register({
            onRequest(request) {
                console.log('onRequest');
                request.headers = {
                    ...request.headers,
                    'X-Client': 'Elsa Dashboard',
                    Authorization: 'Bearer secret-token' // For example, set a JWT value here.
                };
                return request;
            },

            onResponse(response) {
                console.log('onResponse');
                return response;
            }
        });

    })
</script>
```

Or even better: implement your custom middleware as a plugin:

```html
<elsa-studio-root server-url="https://localhost:11000" monaco-lib-path="build/assets/js/monaco-editor/min" culture="en-US">
    <elsa-workflow-definition-editor-screen></elsa-workflow-definition-editor-screen>
</elsa-studio-root>

<script>
    // Custom plugin that intercepts outgoing HTTP requests and their responses.
    function CustomHttpMiddlewarePlugin(elsaStudio) {
        const eventBus = elsaStudio.eventBus;

        eventBus.on('http-client-created', e => {
            // Register a sample middleware.
            e.service.register({
                onRequest(request) {
                    console.log('onRequest');
                    return request;
                },

                onResponse(response) {
                    console.log('onResponse');
                    return response;
                }
            });
        });
    }

    // Get a handle to the elsa-studio-root element.
    const elsaStudioRoot = document.querySelector('elsa-studio-root');

    // Configure Elsa.
    elsaStudioRoot.addEventListener('initializing', e => {
        const elsa = e.detail;
        elsa.pluginManager.registerPlugins([CustomReadLinePlugin, CustomBulkActionsPlugin, CustomHttpMiddlewarePlugin]);
    });
</script>
```

**Reply by apnoda** [reply]
@sfmskywalker 
This is not working anymore in the 2.4.1-preview.11. 
I think the issue is in the elsa-studio-root.tsx, where calling `const elsaClient = await elsaClientFactory();` emits the EventTypes.HttpClientCreated, before EventTypes.Root.Initializing is emitted with the `await eventBus.emit(EventTypes.Root.Initializing);`

**Reply by sfmskywalker** [reply]
You're right, see #2653
Should be fixed as of `2.4.1-preview.13` 👍🏻
